### ISSUE TITLE
Improve performance and fix scalability issue with matching Analytics properties by tag or site URL

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3012]
There are two scenarios in the Site Kit codebase where a significant amount of Analytics API requests may be necessary to match a property in _any_ of the user's accounts based on certain requirements:
* The `GET:accounts-properties-profiles` datapoint relies on the logic in `GET:properties-profiles` to find a matching property by website URL, but it calls that datapoint for every account until something is found (or not), i.e. this process results in a ton of API requests to get properties for different accounts. We even have put a limit of 25 in place to prevent quota errors or timeouts.
* The `GET:tag-permission` datapoint behaves fairly similar, relying on `GET:properties-profiles` until it gets to a list of properties which includes the requested property ID, which results in the same issue. Worse actually, there is no limit in place here, so this could run into quota issues.

**TLDR: The number of requests to match a property by site URL or property ID across all accounts is bonkers.**

Turns out, there is an [accountSummaries/list API endpoint](https://developers.google.com/analytics/devguides/config/mgmt/v3/mgmtReference/management/accountSummaries/list) which allows us to prevent this overload of request (🤦‍♂️  for not having known about this before). The response of that endpoint contains a "summary" of all accounts which includes fewer details **but all of the nested properties and profiles** with enough information to solve our concerns above.

Let's update our codebase to use this endpoint.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The `GET:accounts-properties-profiles` datapoint should rely on [accountSummaries/list](https://developers.google.com/analytics/devguides/config/mgmt/v3/mgmtReference/management/accountSummaries/list) to get a list of all properties between all accounts instead of calling `GET:properties-profiles` for each account.
    * The functionality to match a property based on website URL which is currently within `GET:properties-profiles` should be moved to a reusable method so that it can also be used in `GET:accounts-properties-profiles`.
    * The response shape should not be affected, and the returned data should still rely on **full** account, property and profile objects. In other words, only use the account summary objects for the matching, but once a match is found, rely on a one request to get the full properties for the matched account and another request to get the full profiles for the matched property.
    * The limit of checking only for 25 accounts here is now obviously unnecessary and can be removed.
* The `GET:tag-permission` datapoint (`has_access_to_property` method) should also rely on [accountSummaries/list](https://developers.google.com/analytics/devguides/config/mgmt/v3/mgmtReference/management/accountSummaries/list) to get a list of all properties between all accounts instead of calling `GET:properties-profiles` for each account.
    * This should be the only request needed for the entire datapoint here, because it doesn't return full objects, but only the relevant IDs, which is already present in the account summaries. So this datapoint shouldn't even use `accounts-properties-profiles` anymore.
* The behavior of `GET:properties-profiles` should be unchanged (since it only uses data from a single properties request anyway to find a match).

## Implementation Brief

* Add a new method `Analytics::find_property( array $properties, $property_id = '', array $urls = array() )`. This method should loop through incoming properties array and find a property which has either an ID that equals to `$property_id` or a websiteUrl that equals to one of the provided URLs if `$urls` array is not empty. Use `$property->getId()` and `$property->getWebsiteUrl()` method to get property settings.
  * If the incoming `$property_id` argument is empty, then grab the current `propertyID` from the module settings as it's done [here](https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L847-L848).
* Update the [fallback section](https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L784-L804) of the `GET:accounts-properties-profiles` case in the `Analytics::parse_data_response` method to do the following instead of the current functionality:
  * Pull all account summaries for the current user using the following call:
    ```php
    $account_summaries = $this->get_service( 'analytics' )->management_accountSummaries->listManagementAccountSummaries();
    ```
  * Loop through each account summaries and call the `find_property( $account->getWebProperties(), '', $this->permute_site_url( $current_url ) )` method.
  * If we get a property for one of the account summaries, then break the loop and use that account ID to get full information about properties and profiles for that account using the following snippet:
    ```php
    $properties_profiles = $this->get_data( 'properties-profiles', array( 'accountID' => $account->getId() ) );
    ```
* Update the `GET:properties-profiles` case in the `Analytics::parse_data_response` method to use the new `find_property` method instead of the following section: https://github.com/google/site-kit-wp/blob/55c3280c1e61cbce8aa9360592c72a3fe452460b/includes/Modules/Analytics.php#L840-L865
  * Call `$found_property = $this->find_property( $properties, $data['existingPropertyID'], array() )` if the `$data['existingPropertyID']` is not empty;
  * Otherwise call `$found_property = $this->find_property( $properties, '', $this->permute_site_url( $current_url ) )`.
* Update the `Analytics::has_access_to_property` method to use the same `management_accountSummaries->listManagementAccountSummaries()` function call to retrieve all accounts, properties, and profiles for the current user and use it instead of sending requests to `properties-profiles` and `accounts-properties-profiles` endpoints to find whether the current user has to property or not.

### Test Coverage

* Update the existing tests if any are broken due to the aforementioned changes.

### Visual Regression Changes

* N/A

## QA Brief

* Go to the Analytics module settings and make sure you see the same list of accounts, properties and profiles as you can see if you install the current latest version of the plugin.

## Changelog entry

* Improve speed and reliability of Analytics property matching.

**Comment by felixarntz** [comment]
cc @aaemnnosttv This is a fairly big issue now, but I think it makes sense to handle it together as the implementation may get some overlap, e.g. reusable private methods.

**Comment by eugene-manuilov** [comment]
@felixarntz this is ready for IB review. The only think that I am not sure about is the following statement in the AC section:

> The functionality to match a property based on website URL which is currently within GET:properties-profiles should be moved to a reusable method so that it can also be used in GET:accounts-properties-profiles.

Not sure why we need to do it because we don't use it in the handler for `GET:accounts-properties-profiles` endpoint.

**Comment by felixarntz** [comment]
@eugene-manuilov I think the IB here is a bit off.

> Use this `$accounts_summary` in the fallback section (see [here](https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L784-L804)) to find intersections with accounts in the `$accounts` array and use the first intersected account to make a request to the `properties-profiles` endpoint (using `$this->get_data( 'properties-profiles', array( 'accountID' => $account->getId() ) )`).

What do you mean by intersecting accounts? I think this misses one of the main points in the ACs which is to _not_ use `GET:properties-profiles` to match the _property_. Basically we can assume that all accounts in the `$account_summaries` also exist in `$accounts`. The reason for using `$account_summaries` is that in there it contains all properties for all accounts, so we'll need to iterate over them to find the matching property (_that_ part is something that is currently only done by `GET:properties-profiles`, but we shouldn't call it for that purpose anymore because it also makes a (now unnecessary) API call for each account.

Once we've found the matching property (if there is any), then we should request the full list of profiles for that property, as well as the full list of properties for the account that the matched property is part of (so that we return similar data like we do today). If there's no matching property, we can just return empty arrays for properties and profiles (also similar to how it is today).

> > The functionality to match a property based on website URL which is currently within GET:properties-profiles should be moved to a reusable method so that it can also be used in GET:accounts-properties-profiles.
>
> Not sure why we need to do it because we don't use it in the handler for `GET:accounts-properties-profiles` endpoint.

That relates to the above - we don't use it there now, but we _should_ have the matching logic in there now because it should be done based on properties within `$account_summaries` instead of properties from `GET:properties-profiles`.

**Comment by felixarntz** [comment]
@eugene-manuilov IB is pretty much good to go now! I just have one concern:

> * Add a new method `Analytics::find_property( array $properties, $property_id = null, $urls = null )`. This method should loop through incoming properties array and find a property which has either an ID that equals to `$property_id` or a websiteUrl that equals to one of the provided URLs. Use `$property->getId()` and `$property->getWebsiteUrl()` method to get property settings.
>    * If the incoming `$property_id` argument is null, then grab the current `propertyID` from the module settings as it's done [here](https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L847-L848).
>    * If the incoming `$urls` argument is not an array, then get the current URL using the `$this->context->get_reference_site_url()` method call and get current urls using the `$this->permute_site_url( $current_url )` method call as it's done [here](https://github.com/google/site-kit-wp/blob/develop/includes/Modules/Analytics.php#L849).

I think using these fallbacks here is rather error-prone. I'd prefer if we expected the callers to pass these values, otherwise I can easily see some odd behavior here since it would be necessary to know about these fallbacks/defaults when using the method. For example, for the `has_access_to_property` method, the URLs are completely irrelevant - why should `find_property` still try to rely on the current site URL there? Passing an empty array to disable that behavior feels counter-intuitive to me.

I suggest we don't use any defaults/fallbacks and pass them instead as needed. The function itself should only try to match the property ID if one is passed, and it should only try to match a URL if at least one is passed. We can then also simplify the parameter types like `find_property( array $properties, string $property_id = '', array $urls = array() )` (not really important, but it won't be needed to check for `null` anymore, and it's a bit cleaner IMO).

**Comment by eugene-manuilov** [comment]
@felixarntz ok, how about now? I have updated IB to use your suggestion.

**Comment by felixarntz** [comment]
@eugene-manuilov Lgtm, except I think there is one remainder of your previous IB here that needs to be fixed:

> Loop through each account summaries and call the `find_property( $account->getWebProperties(), '', array() )` method.

I think this also needs to passed the permuted site URL variants now that this is not an automatic fallback of the method.

**Comment by eugene-manuilov** [comment]
Ah.. yes... Good catch! Thanks! Updated.

**Comment by felixarntz** [comment]
IB ✅

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

**Verified:**

- Analytics module settings display the same list of accounts, properties and profiles between `develop` and the `latest release`.   Latest release - [Screenshot](https://user-images.githubusercontent.com/73545194/115026080-a4dba900-9ef4-11eb-9fe8-8cacfb150856.png) & Develop branch - [Screenshot](https://user-images.githubusercontent.com/73545194/115026098-ab6a2080-9ef4-11eb-9308-6a11a71787b5.png)
